### DISCUSSION TITLE
How to generate a `cue.Value` from a struct{}

### DISCUSSION THREAD

**Discussion by Dentrax** [cue-lang/cue#1252] [Q&A]
Let's assume we have the following struct:

```golang
type foo struct {
 bar int
}
```

What we want to achieve is that we simple generate a `cue.Value` from the _initialized_ struct:

```golang
myCue := foo{
  bar: 7,
}

// myCueValue is a cue.Value
// FromGoType is a representative func
myCueValue := cue.FromGoType(myCue)
```

Which will convert into:

```cue
package lookup

foo: {
  bar: 7
}
``` 

Then, we will be able to validate with our _custom cue.Value_ as follows:

```golang
import "cuelang.org/go/encoding/json"

bad, _ := ioutil.ReadFile("bad.json")
err := json.Validate(bad, myCueValue)
```

Not sure if it would solve our use-case but we found [FromGoValue()](https://github.com/cue-lang/cue/blob/0f53054da7243c7a312cd0ad54a1083d67c185d3/internal/value/value.go#L77-L84) and [FromGoType](https://github.com/cue-lang/cue/blob/0f53054da7243c7a312cd0ad54a1083d67c185d3/internal/value/value.go#L86-L97) functions. But these are in the `internal` package intentionally, I guess. @mpvl, right?

Any thoughts? Thanks!

@developer-guy @erkanzileli

**Comment by verdverm** [comment]
You should be able to pass the Go struct to `value.FillPath` based on [this line](https://github.com/cue-lang/cue/blob/v0.4.0/cue/types.go#L1660). I often use `ParsePath` as the first argument to `FillPath`.

- https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#Value.FillPath
- https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#ParsePath

**Reply by Dentrax** [reply]
Thanks! I wrote some minimal reproducible Go snippet for this case. But following code throws _panic_. Any ideas? 🤔 

```golang
func main() {
	type foo struct { bar int }
	root := cue.ParsePath("gen.json")
	value := cue.Value{}.FillPath(root, foo{})
	b, _ := ioutil.ReadFile("bad.json")
	_ = json.Validate(b, value)
}
```

**gen.json**:
```json
{
  "foo": {
    "bar": 7
  }
}
```

**bad.json**:
```json
{
  "foo": {
    "bar": 8
  }
}
```

**panic**:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x126dee2]

goroutine 1 [running]:
cuelang.org/go/internal/core/runtime.(*Runtime).Init(...)
        /Users/furkan.turkal/go/src/cue-demo/vendor/cuelang.org/go/internal/core/runtime/runtime.go:46
cuelang.org/go/cue.(*Runtime).runtime(...)
        /Users/furkan.turkal/go/src/cue-demo/vendor/cuelang.org/go/cue/build.go:39
cuelang.org/go/cue.(*Runtime).Compile(0x0, {0x1531ba1, 0x200}, {0x1494760, 0xc0002f2c90})
        /Users/furkan.turkal/go/src/cue-demo/vendor/cuelang.org/go/cue/build.go:63 +0x62
cuelang.org/go/pkg/encoding/json.Validate({0xc0002b2200, 0x1f, 0x200}, {0x0, 0x0, 0x0})
        /Users/furkan.turkal/go/src/cue-demo/vendor/cuelang.org/go/pkg/encoding/json/manual.go:115 +0x89
cuelang.org/go/encoding/json.Validate(...)
        /Users/furkan.turkal/go/src/cue-demo/vendor/cuelang.org/go/encoding/json/json.go:41
main.main()
        /Users/furkan.turkal/go/src/cue-demo/main.go:25 +0xa8
```

**Reply by verdverm** [reply]
I suspect you need to create an empty value with something like: https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#Context.CompileString, which itself will require making a new context

**Reply by Dentrax** [reply]
Thank you so much @verdverm! 🤗 If you wish, we can add this demo as a Go example in documentation at the section called [Processing CUE in Go](https://cuelang.org/docs/integrations/go/#processing-cue-in-go). Section name would be something like: _Generating CUE Values and Validating_. What do you think?

The following snippet working fine, I'm dropping here, so it could give a help for people who interested in!

```golang
package main

import (
	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/encoding/json"
	"io/ioutil"
)

func main() {
	input := `{
  "foo": {
    "bar": 8
  }
}`
	ctx := cuecontext.New()
	v := ctx.CompileString(input, []cue.BuildOption{}...)
	b, _ := ioutil.ReadFile("policy.cue")
	err := json.Validate(b, v)
	if err != nil {
		panic(err)
	}
}
```

**Policy:**

```cue
{
  "foo": {
    "bar": 7
  }
}
```

**Using:**

* Run the project and expect `exit code 0`
* Set `foo.bar` to `8` in `policy.cue` file and expect: `panic: foo.bar: conflicting values 7 and 8`

**Reply by developer-guy** [reply]
let us introduce the problem that we want to sort out @verdverm, we have a JSON file that we download and we want to accept the policies from the users via flags something like that `-policy policy.cue`, so we thought that we can validate our JSON file against these policies written with Cuelang, does it make sense, or is this code enough to do that, thank you. If you want to learn more about the issue, please [see](https://github.com/sigstore/cosign/issues/512).

**Reply by verdverm** [reply]
@Dentrax I don't manage the docs / website here. I do run Cuetorials and would be happy to add this example here: https://cuetorials.com/api/  Open a PR and we can go from there.

@developer-guy would need to understand more about where this fits in the sigstore process (cli vs server for example). The example above is a bit weird in that a CUE file is validated with the JSON function. Can you expand on this over at the issue you linked? I can reply there, so as to keep this discussion on topic.
